### PR TITLE
Fixes #26424 - refactor autocomplete errors display

### PR DESCRIPTION
**Pull request by Ron-Lavi**
continuing [filters autocomplete PR](https://github.com/theforeman/foreman/pull/6142) breakdown and [discussion](https://github.com/theforeman/foreman/pull/6142#discussion_r242213533).

### CODE REVIEW

**Comment by Ron-Lavi** [comment]
@tbrisker @sharvit @amirfefer could you have a look ? in continues to the [discussion](https://github.com/theforeman/foreman/pull/6142#discussion_r242213533) in filters Autocomplete PR

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 1]
```diff
@@ -26,26 +27,30 @@ export const getResults = ({
...
   trigger,
   id,
   dispatch,
-}) =>
-  API.get(path)
+  url,
+}) => {
+  if (!url) {
+    requestFailure({ error: null, id, dispatch });
+    throw new Error(__('No API path was provided.'));
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: translation  here is redundant

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteConstants.js [hunk 2]
```diff
@@ -2,6 +2,7 @@ export const AUTO_COMPLETE_REQUEST = 'AUTO_COMPLETE_REQUEST';
 export const AUTO_COMPLETE_SUCCESS = 'AUTO_COMPLETE_SUCCESS';
 export const AUTO_COMPLETE_FAILURE = 'AUTO_COMPLETE_FAILURE';
 export const AUTO_COMPLETE_RESET = 'AUTO_COMPLETE_RESET';
+export const ERROR_TO_SHOW = 'ERROR_TO_SHOW';
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteConstants.js]: `ERROR_TO_SHOW` is too ambiguous IMHO

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 3]
```diff
@@ -56,7 +61,17 @@ let createAPIRequest = ({
         id,
       })
     )
-    .catch(error => requestFailure({ error, id, dispatch }));
+    .catch(error => {
+      const { type } = error;
+      if (type !== ERROR_TO_SHOW) {
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: Is this specific only to `Autocomplere`? what about a global error mechanism?

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @Laviro , I left few inline comments

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteConstants.js (hunk 2 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteConstants.js]: sorry I got carried away :sweat_smile:

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js (hunk 3 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: when the autocomplete query is incorrect, an error should return from the server,
an error that we want to show to the user. other than that, all other errors should be visible in the console only. Do we have something similar elsewhere? Thanks

**Comment by Laviro** [comment]
ready for more reviews :)

**Comment by Ron-Lavi** [comment]
@tbrisker @ares could you take a look ? Thanks

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 4]
```diff
@@ -83,7 +96,8 @@ const requestSuccess = ({
 }) => {
   const { error } = data[0] || {};
   if (error) {
-    throw error;
+    // eslint-disable-next-line no-throw-literal
+    throw { message: error, visible: true }; // display this error in the UI.
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: Won't it be easier to dispatch an error instead of `throwing an error -> cathing the error -> verify the error type -> dispatch the error`?

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @Laviro 
I believe it's a good opportunity refactor the actions file. I think it became to massy

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js (hunk 4 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: in this case I am only dispatching the errors which will be visible to the user,
because once the error was passed as a prop, it would be visible..
the errors that I would like to display are only the errors that was return by `success` from the server.
 
another option would be to pass an object similar to the above through the reducer,
and than make this logic in the component - it would keep this mechanism but in my opinion would be a lot messy.

**Comment by Ron-Lavi** [comment]
Thanks @sharvit, we could refactor the action file,
but I ask you to wait with it until I will finish https://github.com/theforeman/foreman/pull/6615 which will bring more changes into this file.

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js (hunk 4 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: Maybe i'm missing something but wouldn't it be the same If you replace the `throw` with:
```js
dispatch(requestFailure({ error, id }));
```

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: gotcha, I thought you were talking about another thing.. :joy:

makes sense, updated :+1:

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 5]
```diff
@@ -26,26 +26,30 @@ export const getResults = ({
...
   trigger,
   id,
   dispatch,
-}) =>
-  API.get(path)
+  url,
+}) => {
+  if (!url) {
+    requestFailure({ error: null, id, dispatch });
+    throw new Error('No API path was provided.');
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: Why do you need to `throw` this error? In other words, who catch and respond to this error?
Shouldn't `return requestFailure({ error: null, id, dispatch });` is good enough?

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: these errors that I `throw` will be visible only in the console with no need to catch or respond to.
the `requestFailure` is being called to indicate that there was a failure,
but the error is not always interesting for the user - as discussed here: https://github.com/theforeman/foreman/pull/6142#discussion_r242213533.

in `requestFailure` I am returning `{ error: null, ...}` so the error won't be visible.

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 6]
```diff
@@ -99,16 +108,17 @@ const requestSuccess = ({
   });
 };
 
-const requestFailure = ({ error, id, dispatch }) =>
+const requestFailure = ({ error, id, dispatch }) => {
   dispatch({
     type: AUTO_COMPLETE_FAILURE,
     payload: {
       results: [],
-      error: error.message || error,
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:107]: why is this change needed? could there be an `error.message` in certain conditions?

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js (hunk 5 above)
Comment (tbrisker) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: That comment was specifically regarding this case, but in the case of network errors i think it actually makes sense to tell the user it failed, rather than the console showing "unhandled promise rejection error".
Also, I think perhaps it makes sense to use `console.warn` in this case instead of throwing errors that aren't caught?

Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: IMO we should never throw errors if we have no intention to catch them as a golden rule, errors should be handled by us and not by the browser.

In redux actions, it doesn't make any sense to throw errors because we update the UI based on state changes and not based on event calls.

In other words:
```js
const MyCounter = ({ count, updateCount }) => {
  const addOne = () => {
    try {
      updateCount(count + 1);
    } catch (error) {
      // There is no much i can do here because I can't render the error from here
      // Instead throwing an error, it would make more sense to
      // update the state so it will cause the UI to get rerendered with error
    }
  };

  return (
    <div>
      <span>{count}</span>
      <button onClick={addOne}>Add one</button>
    </div>
  );
};
```

I agree with @tbrisker, `console.warn` would make more sense here.

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: thanks guys, makes sense..update with your suggestion.. although I prefer to use `console.error` instead of a warning.. 
@tbrisker you think we should display the network error message?
@sharvit, I am going to update the error catching mechanism that you suggested.

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js (hunk 6 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:107]: when I threw the `new Error()` it contained the message property,
but it had resolved now since we are using the `requestFailure` action instead.

**Comment by Ron-Lavi** [comment]
@sharvit, I updated with the `try catch` mechanism you suggested

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @Laviro :+1: 

imo `requestFailure` should always receive a real `error` (`new Error('error message')`) so it will contain a `stacktrace`

**Comment by Ron-Lavi** [comment]
Thanks @sharvit, added `new Error` for each `requestFailure` caller.
@tbrisker, Network errors will be visible while other API errors will remain in the console :+1:

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 7]
```diff
@@ -26,37 +26,62 @@ export const getResults = ({
...
-}) =>
-  API.get(path)
-    .then(({ data }) =>
-      requestSuccess({
-        data,
-        controller,
+  url,
+}) => {
+  if (!url) {
+    requestFailure({
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: should be returned to prevent the API call?

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 8]
```diff
@@ -99,16 +136,21 @@ const requestSuccess = ({
...
+  if (!isVisible) {
+    // eslint-disable-next-line no-console
+    console.error('[Autocomplete]:', error);
+  }
   dispatch({
     type: AUTO_COMPLETE_FAILURE,
     payload: {
       results: [],
-      error: error.message || error,
+      error: isVisible ? error : null,
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: Instead of `console.error`, would it make more sense to dispatch payload of `{ error, isVisible }` so it always logs and stores the error?

A selector can select the error only if it should be visible.

**Review by sharvit — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js (hunk 8 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: thanks, I like it better than using the console

**Comment by Ron-Lavi** [comment]
@sharvit, added return to each `requestFailure` and changed the error from a `string` into an `object` in the `state`

**Comment by Ron-Lavi** [comment]
In the last push, I've added an option to display an `initialError` - error when first mounting the component,
which might be needed in https://github.com/theforeman/foreman/pull/6531
cc @amirfefer

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 9]
```diff
@@ -26,37 +26,61 @@ export const getResults = ({
...
+    return requestSuccess({
+      data,
+      controller,
+      dispatch,
+      searchQuery,
+      trigger,
+      id,
+    });
+  } catch (error) {
+    const catchedError = new Error(error.message);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: Why? The `error` you received is already an `Error`, once you create a new `Error` you will lose the original error stack-trace which is what you need to debug and understand the error.

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 10]
```diff
@@ -26,37 +26,61 @@ export const getResults = ({
...
+    if (catchedError.message === 'Network Error') {
+      return requestFailure({
+        error: catchedError.message,
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: Didn't you want to use the `error` itself instead of the `string` message?

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 11]
```diff
@@ -99,16 +135,17 @@ const requestSuccess = ({
...
 };
 
-const requestFailure = ({ error, id, dispatch }) =>
+const requestFailure = ({ error, id, dispatch, isVisible = true }) => {
   dispatch({
     type: AUTO_COMPLETE_FAILURE,
     payload: {
       results: [],
-      error: error.message || error,
+      error: { message: error, isVisible },
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js]: The `error` is an object so imo it shouldn't be called `message`.
I think `payload: { error, isErrorVisible: isVisible }` represents the values better.

**Review by sharvit — CHANGES_REQUESTED** [review]

**Comment by Ron-Lavi** [comment]
@sharvit thanks for the review, fixed with your suggestions :)

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js [hunk 12]
```diff
@@ -99,16 +126,18 @@ const requestSuccess = ({
...
 };
 
-const requestFailure = ({ error, id, dispatch }) =>
+const requestFailure = ({ error, id, dispatch, isVisible = true }) => {
   dispatch({
     type: AUTO_COMPLETE_FAILURE,
     payload: {
       results: [],
-      error: error.message || error,
+      error: error.message,
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:134]: Can you put the full `error` object in the state? We might want to read `error.stack` as well when debugging errors.

**Review by sharvit — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js (hunk 12 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:134]: can use selector to get the error message from the state

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/AutoComplete/AutoCompleteActions.js:134]: for some reason, passing the error object to the store finish with an empty object, that's why I used the `error.message`, I guess the problem is that the Error object cannot be serialized.

**Review by sharvit — APPROVED** [review]
Thanks @Laviro LGTM :+1:

**Review by tbrisker — APPROVED** [review]
Thanks @Laviro and @sharvit !
